`chainer.grad` is broken with ChainerX backend

As reported by @hvy, `chainerx.grad` called by `chainer.grad` seems broken with macOS clang.  https://travis-ci.org/chainer/chainer/builds/577696208 
```
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_0_{extend_graph_x=False, extend_graph_y=False}::test_double_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_0_{extend_graph_x=False, extend_graph_y=False}::test_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_2_{extend_graph_x=True, extend_graph_y=False}::test_double_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_2_{extend_graph_x=True, extend_graph_y=False}::test_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_1_{extend_graph_x=False, extend_graph_y=True}::test_double_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_1_{extend_graph_x=False, extend_graph_y=True}::test_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_3_{extend_graph_x=True, extend_graph_y=True}::test_double_grad
FAIL tests/chainer_tests/test_function_node.py::TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_3_{extend_graph_x=True, extend_graph_y=True}::test_grad
```
```
_ TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_0_{extend_graph_x=False, extend_graph_y=False}.test_grad _
self = <chainer.testing._bundle.TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None...t_batch_normalization_false__use_ideep_never_param_0_{extend_graph_x=False, extend_graph_y=False} testMethod=test_grad>
backend_config = <BackendConfig use_chainerx=True chainerx_device='native:0' use_cuda=False cuda_device=None use_cudnn='never' cudnn_deterministic=False autotune=False cudnn_fast_batch_normalization=False use_ideep='never'>
    def test_grad(self, backend_config):
        if self.loss_scale and backend_config.xp is chainerx:
            pytest.skip('chainerx.grad does not support loss_scale')
        self.use_device(backend_config.device)
>       self.check_grad()
tests/chainer_tests/test_function_node.py:970: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/chainer_tests/test_function_node.py:959: in check_grad
    testing.assert_allclose(self._get_value(a), self._get_value(e))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
x = array([  6., -14., -32.], dtype=float32), y = array([  2.,  -4., -48.], dtype=float32), atol = 1e-05, rtol = 0.0001
verbose = True
    def assert_allclose(x, y, atol=1e-5, rtol=1e-4, verbose=True):
        """Asserts if some corresponding element of x and y differs too much.
    
        This function can handle both CPU and GPU arrays simultaneously.
    
        Args:
            x: Left-hand-side array.
            y: Right-hand-side array.
            atol (float): Absolute tolerance.
            rtol (float): Relative tolerance.
            verbose (bool): If ``True``, it outputs verbose messages on error.
    
        """
        x = backend.CpuDevice().send(utils.force_array(x))
        y = backend.CpuDevice().send(utils.force_array(y))
        try:
            numpy.testing.assert_allclose(
                x, y, atol=atol, rtol=rtol, verbose=verbose)
        except AssertionError as e:
            f = six.StringIO()
            f.write(str(e) + '\n\n')
            f.write(
                'assert_allclose failed: \n' +
                '  shape: {} {}\n'.format(x.shape, y.shape) +
                '  dtype: {} {}\n'.format(x.dtype, y.dtype))
            if x.shape == y.shape:
                xx = numpy.atleast_1d(x)
                yy = numpy.atleast_1d(y)
                err = numpy.abs(xx - yy)
    
                tol_rtol = rtol * numpy.abs(yy).astype(numpy.float64)
                tol_err = atol + tol_rtol
    
                i = numpy.unravel_index(
                    numpy.argmax(err.astype(numpy.float64) - tol_err), err.shape)
    
                if yy[i] == 0:
                    rel_err = 'inf'
                else:
                    rel_err = err[i] / numpy.abs(yy[i])
    
                f.write(
                    '  i: {}\n'.format(i) +
                    '  x[i]: {}\n'.format(xx[i]) +
                    '  y[i]: {}\n'.format(yy[i]) +
                    '  relative error[i]: {}\n'.format(rel_err) +
                    '  absolute error[i]: {}\n'.format(err[i]) +
                    '  relative tolerance * |y[i]|: {}\n'.format(tol_rtol[i]) +
                    '  absolute tolerance: {}\n'.format(atol) +
                    '  total tolerance: {}\n'.format(tol_err[i]))
    
            opts = numpy.get_printoptions()
            try:
                numpy.set_printoptions(threshold=10000)
                f.write('x: ' + numpy.array2string(x, prefix='x: ') + '\n')
                f.write('y: ' + numpy.array2string(y, prefix='y: ') + '\n')
            finally:
                numpy.set_printoptions(**opts)
>           raise AssertionError(f.getvalue())
E           AssertionError: Parameterized test failed.
E           
E           Base test method: TestGradComplex_use_chainerx_true__chainerx_device_native:0__use_cuda_false__cuda_device_None__use_cudnn_never__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never.test_grad
E           Test parameters:
E             extend_graph_x: False
E             extend_graph_y: False
E           
E           
E           (caused by)
E           AssertionError: 
E           Not equal to tolerance rtol=0.0001, atol=1e-05
E           
E           Mismatch: 100%
E           Max absolute difference: 16.
E           Max relative difference: 2.5
E            x: array([  6., -14., -32.], dtype=float32)
E            y: array([  2.,  -4., -48.], dtype=float32)
E           
E           assert_allclose failed: 
E             shape: (3,) (3,)
E             dtype: float32 float32
E             i: (2,)
E             x[i]: -32.0
E             y[i]: -48.0
E             relative error[i]: 0.3333333432674408
E             absolute error[i]: 16.0
E             relative tolerance * |y[i]|: 0.0048000000000000004
E             absolute tolerance: 1e-05
E             total tolerance: 0.00481
E           x: [  6. -14. -32.]
E           y: [  2.  -4. -48.]
../../../.pyenv/versions/3.5.1/lib/python3.5/site-packages/chainer/testing/array.py:68: AssertionError
------------------------------------------------- Captured stdout call -------------------------------------------------
xs  : array([ 1.,  1., -1.], shape=(3,), dtype=float32, device='native:0'), array([ 2.,  3., -2.], shape=(3,), dtype=float32, device='native:0')
gxs : array([ 0., -3., -4.], shape=(3,), dtype=float32, device='native:0'), array([-2.,  0., -4.], shape=(3,), dtype=float32, device='native:0')
gys : array([ 2., -3.,  5.], shape=(3,), dtype=float32, device='native:0'), array([-1.,  2.,  4.], shape=(3,), dtype=float32, device='native:0')
gxs (actual)  : array([  6., -14., -32.], shape=(3,), dtype=float32, device='native:0'), array([ 2., -6., -4.], shape=(3,), dtype=float32, device='native:0')
gxs (expected): array([  2.,  -4., -48.], shape=(3,), dtype=float32, device='native:0'), array([ 0., -2., -4.], shape=(3,), dtype=float32, device='native:0')
```

--- a/chainerx_cc/chainerx/backward.cc
+++ b/chainerx_cc/chainerx/backward.cc
@@ -156,47 +156,43 @@ std::unordered_map<OpNode*, std::vector<uint8_t>> CreateSubgraph(
         // "Forward" traverse array node queue towards outputs.
         while (!candidate_input_array_nodes.empty()) {
             ArrayNode* array_node = candidate_input_array_nodes.back();
+            candidate_input_array_nodes.pop_back();
 
             auto it_op_node = forward_op_nodes.find(array_node);
-            if (it_op_node == forward_op_nodes.end()) {
-                // Array node is not mapped. It could be an output of an op node that was not given as output.
-                candidate_input_array_nodes.pop_back();
-                continue;
-            }
+            // A single array node could have been an input to multiple op nodes.
+            while (it_op_node != forward_op_nodes.end()) {
+                OpNode* op_node = it_op_node->second;
+                if (op_node != nullptr) {  // If array node is not an output.
+                    std::vector<uint8_t>& flags = input_required_flags[op_node];
+                    if (flags.empty()) {
+                        flags.resize(op_node->input_array_node_count());
+                    }
 
-            OpNode* op_node = it_op_node->second;
-            if (op_node == nullptr) {
-                candidate_input_array_nodes.pop_back();
-                continue;  // Array node is an output.
-            }
+                    const std::vector<std::shared_ptr<ArrayNode>>& input_array_nodes = op_node->input_array_nodes();
+                    for (size_t i_input = 0; i_input < op_node->input_array_node_count(); ++i_input) {
+                        if (input_array_nodes[i_input].get() == array_node) {
+                            flags[i_input] = static_cast<int8_t>(true);
+                            // Cannot break since the same inputs may appear more than once in an op node.
+                        }
+                    }
 
-            std::vector<uint8_t>& flags = input_required_flags[op_node];
-            if (flags.empty()) {
-                flags.resize(op_node->input_array_node_count());
-            }
-
-            const std::vector<std::shared_ptr<ArrayNode>>& input_array_nodes = op_node->input_array_nodes();
-            for (size_t i_input = 0; i_input < op_node->input_array_node_count(); ++i_input) {
-                if (input_array_nodes[i_input].get() == array_node) {
-                    flags[i_input] = static_cast<int8_t>(true);
-                    // Cannot break since the same inputs may appear more than once in an op node.
-                }
-            }
-
-            for (const absl::optional<std::weak_ptr<ArrayNode>>& output_array_node : op_node->output_array_nodes()) {
-                if (output_array_node.has_value()) {
-                    if (std::shared_ptr<ArrayNode> out = output_array_node->lock()) {
-                        if (PushNodeIfNotSeen(candidate_input_array_nodes, out.get(), seen_array_nodes)) {
-                            candidate_input_array_nodes_keep_alive.emplace_back(std::move(out));
+                    for (const absl::optional<std::weak_ptr<ArrayNode>>& output_array_node : op_node->output_array_nodes()) {
+                        if (output_array_node.has_value()) {
+                            if (std::shared_ptr<ArrayNode> out = output_array_node->lock()) {
+                                if (PushNodeIfNotSeen(candidate_input_array_nodes, out.get(), seen_array_nodes)) {
+                                    candidate_input_array_nodes_keep_alive.emplace_back(std::move(out));
+                                }
+                            }
                         }
                     }
                 }
-            }
+                forward_op_nodes.erase(it_op_node);
 
-            forward_op_nodes.erase(it_op_node);
+                // Find the next op node that has this array node as input.
+                it_op_node = forward_op_nodes.find(array_node);
+            }
         }
     }
-
     return input_required_flags;
 }
 

